### ISSUE TITLE
Allow for floating references for currently developed libraries in the lock file.

### ISSUE THREAD

**Issue by stefanoborini** [pypa/pipenv#3580]
### Is your feature request related to a problem? Please describe.

When multiple developers are working on an application my-app, and this application is built against a library my-lib in a git repository that is in development (either in a development branch, or on master), the only way to always refer to the latest my-lib is to recreate the lock file every time. 

The reason is that, given the Pipfile containing the reference to my-lib git repo, the lock will contain the current git hashref of the referred branch, which is a fixed ref. 

If multiple developers are working on multiple PRs, each of them will end up recreating and committing the lock file every time, resulting in tons of conflicts over this file. Some of these conflicts emerge from the intrinsic change of my-lib master (or development) branch as it is being developed, leading to a change in the hash ref. Other conflicts may arise from upgrades on the pypi repository that will get into the newly created lockfile as a need to keep synced against my-lib master, if the versions are not pinned in the Pipfile.

### Describe the solution you'd like

requirements.txt is basically a lockfile in disguise if the versions are pinned. However, when presented with a git repo reference, it always retrieves the currently referred branch specified in the (at)branch, floating. This basically kept the changes from the development library coming in without any need to update the requirements.txt all the time.

In my opinion, the lock file should allow for some referenced entries to be kept floating. The build is not deterministic, but during development you don't want that determinism for those entries. Only when it's time to release my-app, when also my-lib is released and pushed maybe to pypi, you want to specify "I want a deterministic binding with my-lib version 1.2".

The idea of pipenv is that you always create the environment from the lock file, using Pipfile as a meta description to create that lockfile, but if the lockfile must change at every PR to follow a developing library, it becomes a constant merge issue, and the environment ends up being created from the Pipfile every time.

### Describe alternatives you've considered

Currently to prevent the conflict effect we use the --skip-lock option and pin the version of the non-floating dependencies in the Pipfile. It works, but I am not sure it's an intended approach. The lack of a development workflow guide introduces the danger of misuse or hacked up solutions.

**Comment by techalchemy** [comment]
At the very least we need better documentation here, but the main idea is that yeah, if your Pipfile points at a VCS repository it's going to get pinned.  It might not make sense for you to be committing lockfiles if that's causing issues and you don't need true reproducibility of your exact execution environment _until you are ready for deployment_.

Really though the only way to ensure you have a current copy of that repository is to re-lock and the only way to know if someone was using the latest version is to see what commit hash they were at (i.e. what was in the lockfile).  And as an extension of that, the only way to reproduce their environment is to do it by commit hash.

We provide one way to work around this -- that is by using `path` identifiers to point at relative paths.  For example:

```
[dev-packages]
vistir = {path = "../vistir", editable = true, extras = ["spinner"]}
```

And I have a copy of `vistir` at the specified location.  Alternatively (and more safely) you can use a subdirectory (a git submodule might make sense).  In both cases you still need additional tooling to manage updating these repositories from the remote 

For the most part, I don't see a lot of people working with things in lockstep like this where an application and a library are so tightly coupled and both are so frequently updated and _also_ merge conflicts cause problems.  My advice is typically -- who cares about merge conflicts in the lockfile, just pick a winner and generate a fresh copy if you are having so much trouble. Alternatively, don't check in your lockfile.  Alternatively, let some automated job generate lockfiles as part of your CI.  I'm not too sure I understand the actual problem here because none of the teams I work on struggle with thhese kinds of issues -- maybe some examples would help?

**Comment by stefanoborini** [comment]
The problem is that requirements.txt specifically allowed this use case. It allowed to download and pin some entities, but leave others editable and free-floating, without the need to recreate it every time with a different hash.

You can't use path because in order to do so, you have to manually change to that path and manually git pull. Besides its inconvenience, it assumes that the path is there, which means that CI build machine will have to perform this operation manually, which is inconvenient.

I am not sure about the solution or the best practice here, and my issue is mostly a point of discussion. I feel like there are two different categories of direct, non-dev dependencies: stable ones and unstable ones. You want to be able to, for the unstable ones, always get the most up to date on git, even when not released.

As you said, I think that the documentation should explain at the very least that this situation exists, and should propose a preferred way to handle this case. What we found out was the following (what we consider) workaround:

1. define strict pinned versions for direct, stable dependencies (e.g. if we depend on click, we specify ``click = "7.0"``, not click="*")
2. define unstable packages as ``unstable-package = { git = 'url', ref = 'master', editable = true }``
3. skip the lock file generation altogether, and build the environment directly from the Pipfile. We do not commit the .lock

I am not 100% sure, but this does not guarantee that indirect dependencies will stay the same. If click 7.0 depends on foo, and foo is 1.2.4, what happens when foo 1.2.5 is released? I suspect the environment will now include 1.2.5. Note that creating the lock file would not solve.

> For the most part, I don't see a lot of people working with things in lockstep like this where an application and a library are so tightly coupled and both are so frequently updated and also merge conflicts cause problems

In both my current and previous employments this is and was a constant occurrence. You have an application and a library, and you modify the library to serve the application. You don't release the library every day, you simply let them both move forward and build the application against the floating master of the library until you release both.

Summing up, the point is that when you develop, you have some stuff that you don't want to change in your environment, and some stuff that you know are in motion. How to solve this dilemma, no idea.

**Comment by matteius** [comment]
@stefanoborini I am wondering if this issue is solvable using the new named package categories feature of pipenv?
